Implement Unsafe opaque and acquire/release put and get methods

Spencer-Comin opened the pull request:
Implement the Unsafe get and put methods with opaque and acquire/release memory semantics by using the corresponding access modes on the unsafe shadow symbols as introduced in https://github.com/eclipse-omr/omr/pull/7517.

Note that this requires a coordinated merge with https://github.com/eclipse-omr/omr/pull/7517.

Spencer-Comin commented:
@hzongaro @zl-wang @0xdaryl @vijaysun-omr FYI

Spencer-Comin commented:
Some performance numbers

## `Unsafe.getInt*()`

<table>
    <tr>
        <th>Arch</th>
        <th>Ordering</th>
        <th>Speedup</th>
    </tr>
    <tr>
        <td rowspan="3">x86</td>
        <td>Acquire</td><td>0.967</td></tr>
    <tr><td>Opaque</td><td>0.988</td></tr>
    <tr><td>Volatile</td><td>0.987</td></tr>
    <tr>
        <td rowspan="3">aarch64</td>
        <td>Acquire</td><td>1.00</td></tr>
    <tr><td>Opaque</td><td>28.2</td></tr>
    <tr><td>Volatile</td><td>0.996</td></tr>
    <tr>
        <td rowspan="3">power</td>
        <td>Acquire</td><td>1.00</td></tr>
    <tr><td>Opaque</td><td>5.99</td></tr>
    <tr><td>Volatile</td><td>1.00</td></tr>
    <tr>
        <td rowspan="3">z</td>
        <td>Acquire</td><td>1.05</td></tr>
    <tr><td>Opaque</td><td>1.05</td></tr>
    <tr><td>Volatile</td><td>1.05</td></tr>
</table>

## `Unsafe.putInt*()`

<table>
    <tr>
        <th>Arch</th>
        <th>Ordering</th>
        <th>Speedup</th>
    </tr>
    <tr>
        <td rowspan="3">x86</td>
        <td>Release</td><td>19.8</td></tr>
    <tr><td>Opaque</td><td>20.8</td></tr>
    <tr><td>Volatile</td><td>1.00</td></tr>
    <tr>
        <td rowspan="3">aarch64</td>
        <td>Release</td><td>1.82</td></tr>
    <tr><td>Opaque</td><td>58.0</td></tr>
    <tr><td>Volatile</td><td>1.00</td></tr>
    <tr>
        <td rowspan="3">power</td>
        <td>Release</td><td>1.91</td></tr>
    <tr><td>Opaque</td><td>57.6</td></tr>
    <tr><td>Volatile</td><td>0.961</td></tr>
    <tr>
        <td rowspan="3">z</td>
        <td>Release</td><td>1.01</td></tr>
    <tr><td>Opaque</td><td>1.01</td></tr>
    <tr><td>Volatile</td><td>1.01</td></tr>
</table>

vijaysun-omr commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -9691,11 +9691,8 @@ bool TR_LocalFlushElimination::examineNode(TR::Node *node, TR::TreeTop *tt, TR::
-               nodeHasSync = true;
-               nodeHasVolatile = true;
-               }
+            nodeHasSync = symReference->getSymbol()->isAcquireRelease();
```
This all looks okay to me, but given that we have had subtle issues with flush elimination, I thought I would tag @klangman for his awareness, and more importantly, if he has some concern.

vijaysun-omr commented:
Looks like a cleaner solution to me since it avoids the `putOrdered` call duplication. Since it still needs reviews from the others, you should get the all-clear before requesting the coordinated merge that appears to be needed.

hzongaro reviewed (commented):
I think these changes look good.  Just a few minor comments and questions.

Spencer-Comin commented on runtime/compiler/codegen/J9CodeGenerator.cpp:
```diff
@@ -1051,7 +1039,7 @@ J9::CodeGenerator::lowerTreeIfNeeded(
       TR::SymbolReference *symRef = node->getSymbolReference();
       TR::Symbol *symbol = symRef->getSymbol();
-      if (symbol->isVolatile() && node->getDataType() == TR::Int64 && !symRef->isUnresolved() && self()->comp()->target().is32Bit() &&
+      if (symbol->isAtLeastOrStrongerThanOpaque() && node->getDataType() == TR::Int64 && !symRef->isUnresolved() && self()->comp()->target().is32Bit() &&
```
I used the `isAtLeastOrStrongerThanOpaque()` form here since this is checking if bitwise atomicity is required, which is necessary for `Opaque` or stronger. I'll add a comment here to clarify that.

Spencer-Comin commented on runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:
```diff
@@ -2636,20 +2634,23 @@ J9::ARM64::TreeEvaluator::flushEvaluator(TR::Node *node, TR::CodeGenerator *cg)
                node = it.currentNode();
-               if (node->getOpCode().hasSymbolReference() && !node->hasUnresolvedSymbolReference() && node->getSymbolReference()->getSymbol()->isVolatile() && !node->getRegister())
+               if (node->getOpCode().hasSymbolReference() &&
+                   !node->hasUnresolvedSymbolReference() &&
+                   node->getSymbolReference()->getSymbol()->isAtLeastOrStrongerThanAcquireRelease()
```
Added a comment in the [last push](https://github.com/eclipse-openj9/openj9/compare/2320726255d923ab2acdad274064c1423b892d22..653ecb2ae5bb614dee5ec4ed1bd46c8102dd88e4)

Spencer-Comin commented on runtime/compiler/p/codegen/J9TreeEvaluator.cpp:
```diff
@@ -3352,20 +3347,23 @@ TR::Register *J9::Power::TreeEvaluator::flushEvaluator(TR::Node *node, TR::CodeG
             {
-            // Iterate the next tree to see if there is a resolved volatile load/store node that has yet to be evaluated
-            // An unresolved volatile might not actually be a volatile access, and therefore can not replace the AllocationFence
+            // Iterate the next tree to see if there is a resolved acquire/release load/store node that has yet to be evaluated
+            // An unresolved acquire/release might not actually be an acquire/release access, and therefore can not replace the AllocationFence
```
Added a comment in the [last push](https://github.com/eclipse-openj9/openj9/compare/2320726255d923ab2acdad274064c1423b892d22..653ecb2ae5bb614dee5ec4ed1bd46c8102dd88e4)

hzongaro reviewed (approved):
Looks good.  Thanks!

0xdaryl commented:
Jenkins test sanity all jdk21 depends eclipse/omr#7517

0xdaryl commented:
Jenkins test sanity all jdk21 depends eclipse/omr#7517

Trying again now that the OMR change is clean.

Spencer-Comin commented:
@0xdaryl the test build failures look like time outs, but giving my changes another look I see some of the same syntax that XLC didn't like in the z/OS and AIX OMR tests. I'll push the same kind of fix, then we could retry testing?

0xdaryl commented:
Jenkins test sanity aix,zos,plinux jdk21 depends eclipse/omr#7517

0xdaryl commented:
Jenkins test sanity all jdk21 depends eclipse/omr#7517

0xdaryl commented:
Jenkins test sanity all jdk21 depends eclipse/omr#7517

0xdaryl commented:
There are still numerous testing timeouts across the board (I believe that is 3 test runs now across different days/times).  Can you verify whether these timeouts are occurring in any other JDK 21 testing, say by inspecting other recent PRs with CI testing, or by creating a dummy PR without your changes that I can launch a subset of these tests on?

0xdaryl commented:
Testing on the dummy PR on the four Linux platforms was 100% clean.  I suggest you look at the timeouts in this PR as somehow related to your changes.

Spencer-Comin commented:
Found the cause of the timeouts. In refactoring I had introduced a typo in one of the uses of the `MemoryOrdering` enum, causing `Unsafe.put*Volatile` and `Unsafe.get*Volatile` calls to generate transparent unsafe symbols rather than volatile unsafe symbols. I've fixed the typo and changed the `MemoryOrdering` enum to an enum class to catch this type of error at compile time. Internal testing (https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/26975/ http://vmfarm.rtp.raleigh.ibm.com/build_info.php?build_id=89980) looks clean apart from a couple failures that look to be either unrelated or build failures.

0xdaryl commented:
Jenkins test sanity all jdk21 depends eclipse/omr#7517

hzongaro reviewed (approved):
I just did a comparison of the latest version of the changes against the previous version, just to make sure there were no unexpected differences.  Looks good.  Thanks!

Spencer-Comin commented:
zlinux failure looks like https://github.com/eclipse-openj9/openj9/issues/18527
